File: plugins/MyFirstJavaPlugin/src/main/java/com/github/yournamehere/MicGain.java
package com.aliucord.plugins;

import android.content.Context;
import android.media.AudioRecord;

import com.aliucord.annotation.AliucordPlugin;
import com.aliucord.entities.Plugin;
import de.robv.android.xposed.XC_MethodHook;

import java.nio.ByteBuffer;

@AliucordPlugin
public class MicGain extends Plugin {
    
    @Override
    public void start(Context context) throws Throwable {
        // 50dB gain multiplier (Increases volume by 316 times!)
        final float multiplier = 316.22f;

        // Hack #1: Intercept standard audio waves
        patcher.patch(AudioRecord.class.getDeclaredMethod("read", short[].class, int.class, int.class), new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                int readSize = (int) param.getResult();
                if (readSize > 0) {
                    short[] audioData = (short[]) param.args[0];
                    int offset = (int) param.args[1];
                    for (int i = offset; i < offset + readSize; i++) {
                        int amplified = (int) (audioData[i] * multiplier);
                        if (amplified > Short.MAX_VALUE) amplified = Short.MAX_VALUE;
                        else if (amplified < Short.MIN_VALUE) amplified = Short.MIN_VALUE;
                        audioData[i] = (short) amplified;
                    }
                }
            }
        });

        // Hack #2: Intercept byte-level audio waves
        patcher.patch(AudioRecord.class.getDeclaredMethod("read", byte[].class, int.class, int.class), new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                int readSize = (int) param.getResult();
                if (readSize > 0) {
                    byte[] audioData = (byte[]) param.args[0];
                    int offset = (int) param.args[1];
                    for (int i = offset; i < offset + readSize - 1; i += 2) {
                        int sample = (audioData[i] & 0xFF) | (audioData[i + 1] << 8);
                        sample = (int) (sample * multiplier);
                        if (sample > Short.MAX_VALUE) sample = Short.MAX_VALUE;
                        else if (sample < Short.MIN_VALUE) sample = Short.MIN_VALUE;
                        audioData[i] = (byte) (sample & 0xFF);
                        audioData[i + 1] = (byte) ((sample >> 8) & 0xFF);
                    }
                }
            }
        });

        // Hack #3: Intercept Discord WebRTC audio streams
        patcher.patch(AudioRecord.class.getDeclaredMethod("read", ByteBuffer.class, int.class), new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                int readSize = (int) param.getResult();
                if (readSize > 0) {
                    ByteBuffer buffer = (ByteBuffer) param.args[0];
                    for (int i = 0; i < readSize - 1; i += 2) {
                        int sample = (buffer.get(i) & 0xFF) | (buffer.get(i + 1) << 8);
                        sample = (int) (sample * multiplier);
                        if (sample > Short.MAX_VALUE) sample = Short.MAX_VALUE;
                        else if (sample < Short.MIN_VALUE) sample = Short.MIN_VALUE;
                        buffer.put(i, (byte) (sample & 0xFF));
                        buffer.put(i + 1, (byte) ((sample >> 8) & 0xFF));
                    }
                }
            }
        });
    }

    @Override
    public void stop(Context context) {
        // Turns off the mic boost when you disable the plugin
        patcher.unpatchAll();
    }
}

File: plugins/MyFirstJavaPlugin/src/main/java/com/github/yournamehere/MicGain.kt
package com.aliucord.plugins

import android.content.Context
import android.media.AudioRecord
import com.aliucord.annotations.AliucordPlugin
import com.aliucord.entities.Plugin
import com.aliucord.patcher.PostHook
import java.nio.ByteBuffer

@AliucordPlugin
class MicGain : Plugin() {

    override fun start(context: Context) {
        // 50dB gain multiplier
        val multiplier = 316.22f

        // Hack #1: Intercept Short Array audio
        patcher.patch(AudioRecord::class.java.getDeclaredMethod("read", ShortArray::class.java, java.lang.Integer.TYPE, java.lang.Integer.TYPE), PostHook { callFrame ->
            val readSize = callFrame.result as Int
            if (readSize > 0) {
                val audioData = callFrame.args[0] as ShortArray
                val offset = callFrame.args[1] as Int
                for (i in offset until offset + readSize) {
                    var amplified = (audioData[i] * multiplier).toInt()
                    if (amplified > 32767) amplified = 32767
                    else if (amplified < -32768) amplified = -32768
                    audioData[i] = amplified.toShort()
                }
            }
        })

        // Hack #2: Intercept Byte Array audio
        patcher.patch(AudioRecord::class.java.getDeclaredMethod("read", ByteArray::class.java, java.lang.Integer.TYPE, java.lang.Integer.TYPE), PostHook { callFrame ->
            val readSize = callFrame.result as Int
            if (readSize > 0) {
                val audioData = callFrame.args[0] as ByteArray
                val offset = callFrame.args[1] as Int
                var i = offset
                while (i < offset + readSize - 1) {
                    var sample = ((audioData[i].toInt() and 0xFF) or (audioData[i + 1].toInt() shl 8)).toShort().toInt()
                    sample = (sample * multiplier).toInt()
                    if (sample > 32767) sample = 32767
                    else if (sample < -32768) sample = -32768
                    audioData[i] = (sample and 0xFF).toByte()
                    audioData[i + 1] = ((sample shr 8) and 0xFF).toByte()
                    i += 2
                }
            }
        })

        // Hack #3: Intercept Discord WebRTC audio streams
        patcher.patch(AudioRecord::class.java.getDeclaredMethod("read", ByteBuffer::class.java, java.lang.Integer.TYPE), PostHook { callFrame ->
            val readSize = callFrame.result as Int
            if (readSize > 0) {
                val buffer = callFrame.args[0] as ByteBuffer
                var i = 0
                while (i < readSize - 1) {
                    var sample = ((buffer.get(i).toInt() and 0xFF) or (buffer.get(i + 1).toInt() shl 8)).toShort().toInt()
                    sample = (sample * multiplier).toInt()
                    if (sample > 32767) sample = 32767
                    else if (sample < -32768) sample = -32768
                    buffer.put(i, (sample and 0xFF).toByte())
                    buffer.put(i + 1, ((sample shr 8) and 0xFF).toByte())
                    i += 2
                }
            }
        })
    }

    override fun stop(context: Context) {
        patcher.unpatchAll()
    }
}
